velociraptor/catalogue/soap_catalogue.py

```diff
@@ -0,0 +1,266 @@
+
+    def get_value(self, group):
+        if self._value is None:
+            values = [term.get_value(group) for term in self.terms]
+            self._value = self.compute_value(*values)
+        return self._value
+
+    def set_value(self, value, group):
+        self._value = value
+
+    def del_value(self, group):
+        del self._value
+        self._value = None
+
+    def compute_value(self, *values):
+        raise NotImplementedError("This calculation has not been implemented!")
+
+
+class VelocityDispersion(CatalogueDerivedDataset):
+    def compute_value(self, velocity_dispersion_matrix):
+        return np.sqrt(
+            velocity_dispersion_matrix[:, 0]
+            + velocity_dispersion_matrix[:, 1]
+            + velocity_dispersion_matrix[:, 2]
+        )
+
+
+class CatalogueGroup(CatalogueElement):
+
+    elements: List[CatalogueElement]
+
+    def __init__(self, file_name, name, handle):
+        super().__init__(file_name, name)
+
+        self.elements = []
+        self._register_elements(handle)
+        self._register_properties()
+        self._register_extra_properties()
+
+    def _register_elements(self, handle):
+        h5group = handle[self.name] if self.name != "" else handle["/"]
+        for key in h5group.keys():
+            h5obj = h5group[key]
+            if isinstance(h5obj, h5py.Group):
+                self.elements.append(
+                    CatalogueGroup(self.file_name, f"{self.name}/{key}", handle)
+                )
+                dynamically_register_properties(self.elements[-1])
+            elif isinstance(h5obj, h5py.Dataset):
+                self.elements.append(
+                    CatalogueDataset(self.file_name, f"{self.name}/{key}", handle)
+                )
+
+    def __str__(self):
+        return f"CatalogueGroup containing the following elements: {[el.name for el in self.elements]}"
+
+    def _register_properties(self):
+        self.properties = {}
+        for el in self.elements:
+            basename = el.name.split("/")[-1].lower()
+            # attribute names cannot start with a number
+            if basename[0].isnumeric():
+                basename = f"v{basename}"
+            if isinstance(el, CatalogueGroup):
+                setattr(self, basename, el)
+            elif isinstance(el, CatalogueDataset):
+                self.properties[basename] = (
+                    el,
+                    property(el.get_value, el.set_value, el.del_value),
+                )
+
+    def _register_extra_properties(self):
+        try:
+            stellar_velocity_dispersion_matrix = self.properties[
+                "stellarvelocitydispersionmatrix"
+            ][0]
+            self.elements.append(
+                VelocityDispersion(
+                    self.file_name,
+                    "stellarvelocitydispersion",
+                    [stellar_velocity_dispersion_matrix],
+                )
+            )
+            el = self.elements[-1]
```

Comment:
Why this is not done in SOAP: because the 1D velocity dispersion can trivially be derived from the velocity dispersion matrix output by SOAP, and we did not want to store any more data in the SOAP output than strictly necessary.

Why this is done here: because VR did output a 1D velocity dispersion, so this is the only way to make sure we can replace VR with SOAP without changing the pipeline. We could also update the pipeline and turn this into a registration function.

Although I must say that I prefer this way over a registration function, since this way will still use lazy reading.

Reply:
I just worry about the potential debt that this function in particular creates... It is a very odd piece of code to have here. I think a registration function may be best? Are you still using registration functions for other things?

Reply:
We do use registration functions for other things, but the problem is that quantities created by registration functions have a different name in the autoplotter. So that would require a change to the pipeline. I was trying to avoid that at this point.

I also personally think this way of doing things has some potential, since it allows you to define derived quantities without necessarily reading the data and computing them. There are tons of registered quantities in the COLIBRE pipeline that are read and computed but never actually used. But I agree that this might not be the cleanest mechanism to create the derived datasets. I have no time left to come up with a better mechanism, but I know @EvgeniiChaikin was interested in exploring this further.

So my proposal is to keep this for now, but aim to remove/change this mechanism in the future, either by using a registration function and updating the pipeline, or by expanding the use of derived datasets through a cleaner mechanism.

Reply:
Ok, we can do that.